crates/wasi-common/src/wasi.rs

```diff
@@ -2,118 +2,474 @@
-    | __WASI_RIGHTS_PATH_REMOVE_DIRECTORY
-    | __WASI_RIGHTS_POLL_FD_READWRITE;
-pub(crate) const RIGHTS_DIRECTORY_INHERITING: __wasi_rights_t =
-    RIGHTS_DIRECTORY_BASE | RIGHTS_REGULAR_FILE_BASE;
-
-// Operations that apply to regular files.
-pub(crate) const RIGHTS_REGULAR_FILE_BASE: __wasi_rights_t = __WASI_RIGHTS_FD_DATASYNC
-    | __WASI_RIGHTS_FD_READ
-    | __WASI_RIGHTS_FD_SEEK
-    | __WASI_RIGHTS_FD_FDSTAT_SET_FLAGS
-    | __WASI_RIGHTS_FD_SYNC
-    | __WASI_RIGHTS_FD_TELL
-    | __WASI_RIGHTS_FD_WRITE
-    | __WASI_RIGHTS_FD_ADVISE
-    | __WASI_RIGHTS_FD_ALLOCATE
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_FD_FILESTAT_SET_SIZE
-    | __WASI_RIGHTS_FD_FILESTAT_SET_TIMES
-    | __WASI_RIGHTS_POLL_FD_READWRITE;
-pub(crate) const RIGHTS_REGULAR_FILE_INHERITING: __wasi_rights_t = 0;
-
-// Operations that apply to sockets and socket pairs.
-pub(crate) const RIGHTS_SOCKET_BASE: __wasi_rights_t = __WASI_RIGHTS_FD_READ
-    | __WASI_RIGHTS_FD_FDSTAT_SET_FLAGS
-    | __WASI_RIGHTS_FD_WRITE
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_POLL_FD_READWRITE
-    | __WASI_RIGHTS_SOCK_SHUTDOWN;
-pub(crate) const RIGHTS_SOCKET_INHERITING: __wasi_rights_t = RIGHTS_ALL;
-
-// Operations that apply to TTYs.
-pub(crate) const RIGHTS_TTY_BASE: __wasi_rights_t = __WASI_RIGHTS_FD_READ
-    | __WASI_RIGHTS_FD_FDSTAT_SET_FLAGS
-    | __WASI_RIGHTS_FD_WRITE
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_POLL_FD_READWRITE;
-#[allow(unused)]
-pub(crate) const RIGHTS_TTY_INHERITING: __wasi_rights_t = 0;
-
-pub fn whence_to_str(whence: __wasi_whence_t) -> &'static str {
-    match whence {
-        __WASI_WHENCE_CUR => "__WASI_WHENCE_CUR",
-        __WASI_WHENCE_END => "__WASI_WHENCE_END",
-        __WASI_WHENCE_SET => "__WASI_WHENCE_SET",
-        other => panic!("Undefined whence value {:?}", other),
+use crate::WasiCtx;
+
+wiggle::from_witx!({
+    witx: ["crates/wasi-common/wig/WASI/phases/snapshot/witx/wasi_snapshot_preview1.witx"],
+    ctx: WasiCtx,
+});
+
+use crate::hostcalls_impl;
+use wiggle_runtime::{GuestPtr, GuestPtrMut, GuestString};
+
+impl wiggle_runtime::GuestErrorType for types::Errno {
+    type Context = WasiCtx;
+
+    fn success() -> types::Errno {
+        types::Errno::Success
+    }
+
+    fn from_error(e: wiggle_runtime::GuestError, ctx: &mut WasiCtx) -> types::Errno {
+        eprintln!("Guest Error: {:?}", e);
+        types::Errno::Inval
     }
 }
 
-pub const __WASI_DIRCOOKIE_START: __wasi_dircookie_t = 0;
+impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
+    fn args_get(
+        &mut self,
+        _argv: GuestPtrMut<GuestPtrMut<u8>>,
+        _argv_buf: GuestPtrMut<u8>,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("args_get")
+    }
+
+    fn args_sizes_get(&mut self) -> std::result::Result<(types::Size, types::Size), types::Errno> {
+        unimplemented!("args_sizes_get")
+    }
+
+    fn environ_get(
+        &mut self,
```

Comment:
One comment I figured would be worth to bring up is to point out the `&mut self` here. That's a pretty powerful assertion in Rust (uniqueness) which means that this couldn't, for example, be called concurrently or recursively. 

Given the nature of wasm and embedding I might personally lean more towards `&self` for all these methods and the type itself gets to determine whether it uses `RefCell` and such or if it uses mutexes/etc internally.

Reply:
That's an excellent point, thanks for bringing it up Alex! I guess it's especially important since multithreaded wasm is essentially round the corner. I reckon it shouldn't be a major deal-breaker if we change `&mut self` to `&self` in `wiggle` but I'll have to sleep on it first.

Reply:
Given the discussion on https://github.com/kubkon/wiggle/issues/32 it sounds like being able to safely recursively call things is definitely desired, so that'd definitely want this to change to `&self` for that reason. For example if a wasi import calls into some sort of host allocation function, that allocation function may recursively (if incorrectly) call into wasi which would then try to double-borrow the `&mut self`